java/code/src/com/redhat/rhn/manager/action/ActionManager.java

```diff
@@ -2344,4 +2367,86 @@ public static ImageInspectAction scheduleImageInspect(User scheduler, List<Long>
         scheduleForExecution(action, new HashSet<>(sids));
         return action;
     }
+
+    /**
+     * Connect given systems to another proxy.
+     *
+     * @param loggedInUser The current user
+     * @param sysids A list of systems ids
+     * @param proxyId Id of the proxy or 0 for direct connection to SUMA server
+     * @return Returns a list of scheduled action ids
+     *
+     */
+
+    public static List<Long> changeProxy(User loggedInUser, List<Long> sysids, Long proxyId)
+        throws TaskomaticApiException {
+        List<Long> visible = MinionServerFactory.lookupVisibleToUser(loggedInUser)
+                    .map(m -> m.getId()).collect(toList());
+        if (!visible.containsAll(sysids)) {
+            sysids.removeAll(visible);
+            throw new UnsupportedOperationException("Some System not available or not managed with Salt: " + sysids);
+        }
+
+        List<MinionServer> minions = sysids.stream().map(
+            id -> SystemManager.lookupByIdAndUser(id, loggedInUser).asMinionServer().get()).collect(toList());
+
+        List<Long> proxies = minions.stream().filter(m -> m.isProxy()).map(m -> m.getId()).collect(toList());
+        if (!proxies.isEmpty()) {
+            throw new UnsupportedOperationException("Some of the minions are proxies: " + proxies);
+        }
+
+        Optional<Server> proxy = Optional.empty();
+
+        if (proxyId != 0) {
+            proxy = Optional.of(SystemManager.lookupByIdAndUser(proxyId, loggedInUser));
+            proxy.ifPresent(p -> {
+                if (!p.isProxy()) {
+                    throw new UnsupportedOperationException("The system is not a proxy: " + p.getId());
+                }
+            });
+        }
+        Optional<Long> proxyIdOpt = proxy.map(p -> p.getId());
+
+        List<Long> sshIds = minions.stream()
+                            .filter(minion -> ContactMethodUtil.isSSHPushContactMethod(minion.getContactMethod()))
+                            .map(minion -> {
+            // handle SSH minions
+            minion.updateServerPaths(proxyIdOpt);
+            ServerFactory.save(minion);
+
+            MinionPillarManager.INSTANCE.generatePillar(minion);
+            return minion.getId();
+        }).collect(toList());
+
+
+        List<Long> normalIds = minions.stream()
+                               .filter(minion -> !ContactMethodUtil.isSSHPushContactMethod(minion.getContactMethod()))
+                               .map(minion -> minion.getId())
+                               .collect(toList());
+
+        List<Long> ret = new ArrayList<Long>();
+        if (!sshIds.isEmpty()) {
+            // action for SSH minions - update channel configuration
+            Action a = scheduleApplyStates(loggedInUser, sshIds,
+                       Collections.singletonList(ApplyStatesEventMessage.CHANNELS),
+                       new Date());
+            a = ActionFactory.save(a);
+            taskomaticApi.scheduleActionExecution(a);
```

Comment:
this method can throw an exception. This will stop the all process. Is this the goal or if it fails for ssh minions we should try to process the normal minions?

Reply:
I think that the situation when one taskomatic call fails and the other succeeds is quite improbable. If there is any problem, both actions fails.
In any case, a correct recovery would be to fix taskomatic and call changeProxy again with the same arguments.

So I think that returning the error immediately is correct.